dali/pipeline/operator/builtin/external_source.cc

```diff
@@ -20,38 +20,71 @@ namespace dali {
 template<>
 void ExternalSource<CPUBackend>::RunImpl(HostWorkspace &ws) {
   std::list<uptr_tl_type> tensor_list_elm;
+  std::list<uptr_tv_type> tensor_vector_elm;
   std::list<uptr_cuda_event_type> internal_copy_to_storage;
+  ZeroCopyInfo copy_info;
   {
     std::unique_lock<std::mutex> busy_lock(busy_m_);
-    tensor_list_elm = tl_data_.PopFront();
-    internal_copy_to_storage = copy_to_storage_events_.PopFront();
+    copy_info = zero_copy_.front();
+    zero_copy_.pop_front();
+    if (copy_info.is_tensor_vector && copy_info.is_zero_copy) {
+      tensor_vector_elm = tv_data_.PopFront();
+    } else {
+      tensor_list_elm = tl_data_.PopFront();
+    }
+    if (!copy_info.is_zero_copy) {
+      internal_copy_to_storage = copy_to_storage_events_.PopFront();
+    }
   }
-  auto &thread_pool = ws.GetThreadPool();
+  if (copy_info.is_zero_copy && !copy_info.is_tensor_vector) {
+    TensorVector<CPUBackend> &output = ws.template OutputRef<CPUBackend>(0);
+    output.ShareData(tensor_list_elm.front().get());
+    // empty tensor_list_elm
+    tensor_list_elm.front()->Reset();
+    RecycleBuffer(tensor_list_elm);
+  } else {
+    auto &thread_pool = ws.GetThreadPool();
+    // sort by the work size
+    sample_ids_.clear();
+    sample_ids_.reserve(batch_size_);
 
-  // sort by the work size
-  sample_ids_.clear();
-  sample_ids_.reserve(batch_size_);
-  for (int sample_id = 0; sample_id < batch_size_; sample_id++) {
-    sample_ids_.emplace_back(volume(tensor_list_elm.front()->tensor_shape(sample_id)), sample_id);
+    if (copy_info.is_zero_copy && copy_info.is_tensor_vector) {
+      auto shape = tensor_vector_elm.front()->shape();
+      for (int sample_id = 0; sample_id < batch_size_; sample_id++) {
+        sample_ids_.emplace_back(volume(shape[sample_id]), sample_id);
```

Comment:
Can't you share TensorVector with output TensorVector (so it will be non-continuos) -> in such case you don't need copy.

Reply:
I can but maybe it is better to have that continuous...
I don't have any strong opinion. I would welcome the second opinion.

Reply:
Done that way.